internal/agent/runtime/docker.go

```diff
@@ -0,0 +1,204 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+
+	retry "github.com/avast/retry-go"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/go-logr/logr"
+	"github.com/tinkerbell/tink/internal/agent"
+	"github.com/tinkerbell/tink/internal/agent/runtime/internal"
+	"github.com/tinkerbell/tink/internal/agent/workflow"
+	"github.com/tinkerbell/tink/internal/ptr"
+	"k8s.io/apimachinery/pkg/util/rand"
+)
+
+var _ agent.ContainerRuntime = &Docker{}
+
+// Docker is a docker runtime that satisfies agent.ContainerRuntime.
+type Docker struct {
+	log    logr.Logger
+	client *client.Client
+}
+
+// Run satisfies agent.ContainerRuntime.
+func (d *Docker) Run(ctx context.Context, a workflow.Action) error {
+	pullImage := func() error {
+		// We need the image to be available before we can create a container.
+		image, err := d.client.ImagePull(ctx, a.Image, types.ImagePullOptions{})
+		if err != nil {
+			return fmt.Errorf("docker: %w", err)
+		}
+		defer image.Close()
+
+		// Docker requires everything to be read from the images ReadCloser for the image to actually
+		// be pulled. We may want to log image pulls in a circular buffer somewhere for debugability.
+		if _, err = io.Copy(io.Discard, image); err != nil {
+			return fmt.Errorf("docker: %w", err)
+		}
+
+		return nil
+	}
+
+	err := retry.Do(pullImage, retry.Attempts(5), retry.DelayType(retry.BackOffDelay))
+	if err != nil {
+		return err
+	}
+
+	// TODO: Support all the other things on the action such as volumes.
+	cfg := container.Config{
+		Image: a.Image,
+		Env:   toDockerEnv(a.Env),
+	}
+
+	failureFiles, err := internal.NewFailureFiles()
```

Comment:
Coming back to this after a bit the interaction between a runtime implementation, the agent, the eventing system, and the action, etc being run in terms of failures feels difficult to follow with just the code. Even if we had documentation (which is prone to get out of date with the code) it's feeling a bit delicate/complicated. Maybe we don't need all these places to be aware of these files? Can we talk/revisit this?